spring-integration-core/src/test/java/org/springframework/integration/handler/MethodInvokingMessageProcessorTests.java

```diff
@@ -550,6 +550,22 @@ public void optionalHeaders(Optional<String> foo, @Header(value = "foo", require
 		assertEquals("FOO", targetObject.arguments.get("foo2"));
 	}
 
+	@Test
+	public void testPrivateMethod() throws Exception {
+		class Foo {
+
+			@ServiceActivator
+			private String service(String payload) {
+				return payload.toUpperCase();
+			}
+
+		}
+
+		MessagingMethodInvokerHelper helper = new MessagingMethodInvokerHelper(new Foo(), ServiceActivator.class, false);
+
+		assertEquals("FOO", helper.process(new GenericMessage<>("foo")));
```

Comment:
add `assertEquals("BAR", helper.process(new GenericMessage<>("bar")));` to exhibit caching problem (in a debugger).